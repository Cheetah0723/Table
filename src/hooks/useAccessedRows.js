import { useMemo } from 'react'

export default function useAccessedRows ({
  debug, data, columns, subRowsKey,
}) {
  return useMemo(
    () => {
      if (debug) console.info('getAccessedRows')

      // Access the row's data
      const accessRow = (data, i, depth = 0) => {
        // Keep the original reference around
        const original = data

        // Process any subRows
        const subRows = data[subRowsKey]
          ? data[subRowsKey].map((d, i) => accessRow(d, i, depth + 1))
          : undefined

        const row = {
          original,
          index: i,
          subRows,
          depth,
        }

        // Create the cells and values
        row.values = {}
        columns.forEach(column => {
          row.values[column.id] = column.accessor ? column.accessor(data) : undefined
        })

        return row
      }

      // Use the resolved data
      return data.map((d, i) => accessRow(d, i))
    },
    [data, columns]
  )
}
