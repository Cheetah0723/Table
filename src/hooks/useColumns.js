import { useMemo } from 'react'
import PropTypes from 'prop-types'

import { getBy } from '../utils'

const propTypes = {
  // General
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      Cell: PropTypes.any,
      Header: PropTypes.any,
    })
  ),
  defaultColumn: PropTypes.any,
}

// Find the depth of the columns
function findMaxDepth(columns, depth = 0) {
  return columns.reduce((prev, curr) => {
    if (curr.columns) {
      return Math.max(prev, findMaxDepth(curr.columns, depth + 1))
    }
    return depth
  }, 0)
}

function decorateColumn(column, defaultColumn, parent, depth, index) {
  // Apply the defaultColumn
  column = { ...defaultColumn, ...column }

  // First check for string accessor
  let { id, accessor, Header } = column

  if (typeof accessor === 'string') {
    id = id || accessor
    const accessorString = accessor
    accessor = row => getBy(row, accessorString)
  }

  if (!id && typeof Header === 'string') {
    id = Header
  }

  if (!id && column.columns) {
    console.error(column)
    throw new Error('A column ID (or unique "Header" value) is required!')
  }

  if (!id) {
    console.error(column)
    throw new Error('A column ID (or string accessor) is required!')
  }

  column = {
    Header: ({ id }) => id,
    Cell: ({ value }) => value,
    show: true,
    ...column,
    id,
    accessor,
    parent,
    depth,
    index,
  }

  return column
}

// Build the visible columns, headers and flat column list
function decorateColumnTree(columns, defaultColumn, parent, depth = 0) {
  return columns.map((column, columnIndex) => {
    column = decorateColumn(column, defaultColumn, parent, depth, columnIndex)
    if (column.columns) {
      column.columns = decorateColumnTree(
        column.columns,
        defaultColumn,
        column,
        depth + 1
      )
    }
    return column
  })
}

// Build the header groups from the bottom up
function makeHeaderGroups(columns, maxDepth, defaultColumn) {
  const headerGroups = []

  const removeChildColumns = column => {
    delete column.columns
    if (column.parent) {
      removeChildColumns(column.parent)
    }
  }
  columns.forEach(removeChildColumns)

  const buildGroup = (columns, depth = 0) => {
    const headerGroup = {
      headers: [],
    }

    const parentColumns = []

    const hasParents = columns.some(col => col.parent)

    columns.forEach(column => {
      const isFirst = !parentColumns.length
      let latestParentColumn = [...parentColumns].reverse()[0]

      // If the column has a parent, add it if necessary
      if (column.parent) {
        if (isFirst || latestParentColumn.originalID !== column.parent.id) {
          parentColumns.push({
            ...column.parent,
            originalID: column.parent.id,
            id: [column.parent.id, parentColumns.length].join('_'),
          })
        }
      } else if (hasParents) {
        // If other columns have parents, add a place holder if necessary
        const placeholderColumn = decorateColumn(
          {
            originalID: [column.id, 'placeholder', maxDepth - depth].join('_'),
            id: [
              column.id,
              'placeholder',
              maxDepth - depth,
              parentColumns.length,
            ].join('_'),
          },
          defaultColumn
        )
        if (
          isFirst ||
          latestParentColumn.originalID !== placeholderColumn.originalID
        ) {
          parentColumns.push(placeholderColumn)
        }
      }

      // Establish the new columns[] relationship on the parent
      if (column.parent || hasParents) {
        latestParentColumn = [...parentColumns].reverse()[0]
        latestParentColumn.columns = latestParentColumn.columns || []
        if (!latestParentColumn.columns.includes(column)) {
          latestParentColumn.columns.push(column)
        }
      }

      headerGroup.headers.push(column)
    })

    headerGroups.push(headerGroup)

    if (parentColumns.length) {
      buildGroup(parentColumns)
    }
  }

  buildGroup(columns)

  return headerGroups.reverse()
}

export const useColumns = props => {
  const {
    debug,
    columns: userColumns,
    defaultColumn = {},
    state: [{ groupBy }],
  } = props

  PropTypes.checkPropTypes(propTypes, props, 'property', 'useColumns')

  const { columns, headerGroups, headers } = useMemo(() => {
    if (debug) console.info('getColumns')

    // Decorate All the columns
    let columnTree = decorateColumnTree(userColumns, defaultColumn)

    // Get the flat list of all columns
    let columns = flattenBy(columnTree, 'columns')

    columns = [
      ...groupBy.map(g => columns.find(col => col.id === g)),
      ...columns.filter(col => !groupBy.includes(col.id)),
    ]

    // Get headerGroups
    const headerGroups = makeHeaderGroups(
      columns,
      findMaxDepth(columnTree),
      defaultColumn
    )

    const headers = flattenBy(headerGroups, 'headers')

    return {
      columns,
      headerGroups,
      headers,
    }
  }, [debug, defaultColumn, groupBy, userColumns])

  return {
    ...props,
    columns,
    headerGroups,
    headers,
  }

  function flattenBy(columns, childKey) {
    const flatColumns = []

    const recurse = columns => {
      columns.forEach(d => {
        if (!d[childKey]) {
          flatColumns.push(d)
        } else {
          recurse(d[childKey])
        }
      })
    }

    recurse(columns)

    return flatColumns
  }
}
