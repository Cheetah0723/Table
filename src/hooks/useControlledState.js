const mergeStateAndProps = (state, defaults, props) => {
  Object.keys(props).forEach(key => {
    if (typeof props[key] !== 'undefined') {
      state[key] = props[key]
    }
  })

  Object.keys(defaults).forEach(key => {
    if (typeof state[key] === 'undefined') {
      state[key] = defaults[key]
    }
  })

  return state
}

export default function useControlledState ([userState, userSetState], defaults, props) {
  const state = mergeStateAndProps(userState, defaults, props)

  const setState = (updater, type) =>
    userSetState(old => updater(mergeStateAndProps(old, defaults, props)), type)

  return [state, setState]
}
