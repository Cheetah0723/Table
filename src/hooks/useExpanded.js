import { useMemo } from 'react'
import PropTypes from 'prop-types'

import { getBy, getFirstDefined, setBy } from '../utils'
import { addActions, actions } from '../actions'
import { defaultState } from './useTableState'

defaultState.expanded = {}

addActions({
  toggleExpanded: '__toggleExpanded__',
  useExpanded: '__useExpanded__'
})

const propTypes = {
  expandedKey: PropTypes.string,
  paginateSubRows: PropTypes.bool
}

export const useExpanded = props => {
  PropTypes.checkPropTypes(propTypes, props, 'property', 'useExpanded')

  const {
    debug,
    rows,
    expandedKey = 'expanded',
    hooks,
    state: [{ expanded }, setState],
    paginateSubRows = true
  } = props

  const toggleExpandedByPath = (path, set) => {
    return setState(old => {
      const { expanded } = old
      const existing = getBy(expanded, path)
      set = getFirstDefined(set, !existing)
      return {
        ...old,
        expanded: setBy(expanded, path, set)
      }
    }, actions.toggleExpanded)
  }

  hooks.row.push(row => {
    const { path } = row
    row.toggleExpanded = set => toggleExpandedByPath(path, set)
  })

  const expandedRows = useMemo(() => {
    if (debug) console.info('getExpandedRows')

    const expandedRows = []

    // Here we do some mutation, but it's the last stage in the
    // immutable process so this is safe
    const handleRow = (row, index, depth = 0, parentPath = []) => {
      // Compute some final state for the row
      const path = [...parentPath, index]

      row.path = path
      row.depth = depth

      row.isExpanded =
        (row.original && row.original[expandedKey]) || getBy(expanded, path)

      if (paginateSubRows || (!paginateSubRows && row.depth === 0)) {
        expandedRows.push(row)
      }

      if (row.isExpanded && row.subRows && row.subRows.length) {
        row.subRows.forEach((row, i) => handleRow(row, i, depth + 1, path))
      }

      return row
    }

    rows.forEach((row, i) => handleRow(row, i))

    return expandedRows
  }, [debug, rows, expandedKey, expanded, paginateSubRows])

  const expandedDepth = findExpandedDepth(expanded)

  return {
    ...props,
    toggleExpandedByPath,
    expandedDepth,
    rows: expandedRows
  }
}

function findExpandedDepth(obj, depth = 1) {
  return Object.values(obj).reduce((prev, curr) => {
    if (typeof curr === 'object') {
      return Math.max(prev, findExpandedDepth(curr, depth + 1))
    }
    return depth
  }, 0)
}
