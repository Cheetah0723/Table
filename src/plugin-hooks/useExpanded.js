import React from 'react'

import {
  actions,
  makePropGetter,
  expandRows,
  useMountedLayoutEffect,
  useGetLatest,
} from '../utils'
import { useConsumeHookGetter, functionalUpdate } from '../publicUtils'

// Actions
actions.toggleExpanded = 'toggleExpanded'
actions.toggleAllExpanded = 'toggleAllExpanded'
actions.setExpanded = 'setExpanded'
actions.resetExpanded = 'resetExpanded'

export const useExpanded = hooks => {
  hooks.getExpandedToggleProps = [defaultGetExpandedToggleProps]
  hooks.stateReducers.push(reducer)
  hooks.useInstance.push(useInstance)
}

useExpanded.pluginName = 'useExpanded'

const defaultGetExpandedToggleProps = (props, { row }) => [
  props,
  {
    onClick: e => {
      e.persist()
      row.toggleExpanded()
    },
    style: {
      cursor: 'pointer',
    },
    title: 'Toggle Expanded',
  },
]

// Reducer
function reducer(state, action, previousState, instance) {
  if (action.type === actions.init) {
    return {
      expanded: {},
      ...state,
    }
  }

  if (action.type === actions.resetExpanded) {
    return {
      ...state,
      expanded: instance.initialState.expanded || {},
    }
  }

  if (action.type === actions.setExpanded) {
    return {
      ...state,
      expanded: functionalUpdate(action.expanded, state.expanded),
    }
  }

  if (action.type === actions.toggleExpanded) {
    const { id, expanded: setExpanded } = action
    const exists = state.expanded[id]

    const shouldExist =
      typeof setExpanded !== 'undefined' ? setExpanded : !exists

    if (!exists && shouldExist) {
      return {
        ...state,
        expanded: {
          ...state.expanded,
          [id]: true,
        },
      }
    } else if (exists && !shouldExist) {
      const { [id]: _, ...rest } = state.expanded
      return {
        ...state,
        expanded: rest,
      }
    } else {
      return state
    }
  }
}

function useInstance(instance) {
  const {
    data,
    rows,
    manualExpandedKey = 'expanded',
    paginateExpandedRows = true,
    expandSubRows = true,
    hooks,
    autoResetExpanded = true,
    state: { expanded },
    dispatch,
  } = instance

  const getAutoResetExpanded = useGetLatest(autoResetExpanded)

  // Bypass any effects from firing when this changes
  useMountedLayoutEffect(() => {
    if (getAutoResetExpanded()) {
      dispatch({ type: actions.resetExpanded })
    }
  }, [dispatch, data])

  const toggleExpanded = (id, expanded) => {
    dispatch({ type: actions.toggleExpanded, id, expanded })
  }

  // use reference to avoid memory leak in #1608
  const getInstance = useGetLatest(instance)

  const getExpandedTogglePropsHooks = useConsumeHookGetter(
    getInstance().hooks,
    'getExpandedToggleProps'
  )

  hooks.prepareRow.push(row => {
    row.toggleExpanded = set => instance.toggleExpanded(row.id, set)

    row.getExpandedToggleProps = makePropGetter(getExpandedTogglePropsHooks(), {
      instance: getInstance(),
      row,
    })
  })

  const expandedRows = React.useMemo(() => {
    if (paginateExpandedRows) {
      return expandRows(rows, { manualExpandedKey, expanded, expandSubRows })
    }

    return rows
  }, [paginateExpandedRows, rows, manualExpandedKey, expanded, expandSubRows])

  const expandedDepth = React.useMemo(() => findExpandedDepth(expanded), [
    expanded,
  ])

  Object.assign(instance, {
    preExpandedRows: rows,
    expandedRows,
    rows: expandedRows,
    toggleExpanded,
    expandedDepth,
  })
}

function findExpandedDepth(expanded) {
  let maxDepth = 0

  Object.keys(expanded).forEach(id => {
    const splitId = id.split('.')
    maxDepth = Math.max(maxDepth, splitId.length)
  })

  return maxDepth
}
