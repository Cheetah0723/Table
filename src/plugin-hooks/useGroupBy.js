import React from 'react'

import * as aggregations from '../aggregations'
import {
  actions,
  makePropGetter,
  defaultGroupByFn,
  getFirstDefined,
  ensurePluginOrder,
  useMountedLayoutEffect,
  useGetLatest,
} from '../utils'
import { useConsumeHookGetter } from '../publicUtils'

// Actions
actions.resetGroupBy = 'resetGroupBy'
actions.toggleGroupBy = 'toggleGroupBy'

export const useGroupBy = hooks => {
  hooks.getGroupByToggleProps = [defaultGetGroupByToggleProps]
  hooks.stateReducers.push(reducer)
  hooks.flatColumnsDeps.push((deps, { instance }) => [
    ...deps,
    instance.state.groupBy,
  ])
  hooks.flatColumns.push(flatColumns)
  hooks.useInstance.push(useInstance)
}

useGroupBy.pluginName = 'useGroupBy'

const defaultGetGroupByToggleProps = (props, { header }) => [
  props,
  {
    onClick: header.canGroupBy
      ? e => {
          e.persist()
          header.toggleGroupBy()
        }
      : undefined,
    style: {
      cursor: header.canGroupBy ? 'pointer' : undefined,
    },
    title: 'Toggle GroupBy',
  },
]

// Reducer
function reducer(state, action, previousState, instance) {
  if (action.type === actions.init) {
    return {
      groupBy: [],
      ...state,
    }
  }

  if (action.type === actions.resetGroupBy) {
    return {
      ...state,
      groupBy: instance.initialState.groupBy || [],
    }
  }

  if (action.type === actions.toggleGroupBy) {
    const { columnId, toggle } = action

    const resolvedToggle =
      typeof toggle !== 'undefined' ? toggle : !state.groupBy.includes(columnId)

    if (resolvedToggle) {
      return {
        ...state,
        groupBy: [...state.groupBy, columnId],
      }
    }

    return {
      ...state,
      groupBy: state.groupBy.filter(d => d !== columnId),
    }
  }
}

function flatColumns(
  flatColumns,
  {
    instance: {
      state: { groupBy },
    },
  }
) {
  // Sort grouped columns to the start of the column list
  // before the headers are built

  const groupByColumns = groupBy
    .map(g => flatColumns.find(col => col.id === g))
    .filter(col => !!col)
  const nonGroupByColumns = flatColumns.filter(col => !groupBy.includes(col.id))

  flatColumns = [...groupByColumns, ...nonGroupByColumns]

  flatColumns.forEach(column => {
    column.isGrouped = groupBy.includes(column.id)
    column.groupedIndex = groupBy.indexOf(column.id)
  })

  return flatColumns
}

const defaultUserAggregations = {}

function useInstance(instance) {
  const {
    data,
    rows,
    flatRows,
    flatColumns,
    flatHeaders,
    groupByFn = defaultGroupByFn,
    manualGroupBy,
    aggregations: userAggregations = defaultUserAggregations,
    hooks,
    plugins,
    state: { groupBy },
    dispatch,
    autoResetGroupBy = true,
    manaulGroupBy,
    disableGroupBy,
    defaultCanGroupBy,
  } = instance

  ensurePluginOrder(plugins, [], 'useGroupBy', ['useSortBy', 'useExpanded'])

  const getInstance = useGetLatest(instance)

  flatColumns.forEach(column => {
    const {
      accessor,
      defaultGroupBy: defaultColumnGroupBy,
      disableGroupBy: columnDisableGroupBy,
    } = column

    column.canGroupBy = accessor
      ? getFirstDefined(
          columnDisableGroupBy === true ? false : undefined,
          disableGroupBy === true ? false : undefined,
          true
        )
      : getFirstDefined(defaultColumnGroupBy, defaultCanGroupBy, false)

    if (column.canGroupBy) {
      column.toggleGroupBy = () => instance.toggleGroupBy(column.id)
    }

    column.Aggregated = column.Aggregated || column.Cell
  })

  const toggleGroupBy = (columnId, toggle) => {
    dispatch({ type: actions.toggleGroupBy, columnId, toggle })
  }

  const getGroupByTogglePropsHooks = useConsumeHookGetter(
    getInstance().hooks,
    'getGroupByToggleProps'
  )

  flatHeaders.forEach(header => {
    header.getGroupByToggleProps = makePropGetter(
      getGroupByTogglePropsHooks(),
      { instance: getInstance(), header }
    )
  })

  hooks.prepareRow.push(row => {
    row.allCells.forEach(cell => {
      // Grouped cells are in the groupBy and the pivot cell for the row
      cell.isGrouped = cell.column.isGrouped && cell.column.id === row.groupByID
      // Repeated cells are any columns in the groupBy that are not grouped
      cell.isRepeatedValue = !cell.isGrouped && cell.column.isGrouped
      // Aggregated cells are not grouped, not repeated, but still have subRows
      cell.isAggregated =
        !cell.isGrouped && !cell.isRepeatedValue && row.canExpand
    })
  })

  const [groupedRows, groupedFlatRows] = React.useMemo(() => {
    if (manualGroupBy || !groupBy.length) {
      return [rows, flatRows]
    }

    // Ensure that the list of filtered columns exist
    const existingGroupBy = groupBy.filter(g =>
      flatColumns.find(col => col.id === g)
    )

    // Find the columns that can or are aggregating
    // Uses each column to aggregate rows into a single value
    const aggregateRowsToValues = (rows, isAggregated) => {
      const values = {}

      flatColumns.forEach(column => {
        // Don't aggregate columns that are in the groupBy
        if (existingGroupBy.includes(column.id)) {
          values[column.id] = rows[0] ? rows[0].values[column.id] : null
          return
        }

        const columnValues = rows.map(d => d.values[column.id])

        let aggregator = column.aggregate

        if (Array.isArray(aggregator)) {
          if (aggregator.length !== 2) {
            console.info({ column })
            throw new Error(
              `React Table: Complex aggregators must have 2 values, eg. aggregate: ['sum', 'count']. More info above...`
            )
          }
          if (isAggregated) {
            aggregator = aggregator[1]
          } else {
            aggregator = aggregator[0]
          }
        }

        let aggregateFn =
          typeof aggregator === 'function'
            ? aggregator
            : userAggregations[aggregator] || aggregations[aggregator]

        if (aggregateFn) {
          values[column.id] = aggregateFn(columnValues, rows, isAggregated)
        } else if (aggregator) {
          console.info({ column })
          throw new Error(
            `React Table: Invalid aggregate option for column listed above`
          )
        } else {
          values[column.id] = null
        }
      })
      return values
    }

    let groupedFlatRows = []

    // Recursively group the data
    const groupRecursively = (rows, depth = 0, parentId) => {
      // This is the last level, just return the rows
      if (depth === existingGroupBy.length) {
        return rows
      }

      const columnId = existingGroupBy[depth]

      // Group the rows together for this level
      let groupedRows = groupByFn(rows, columnId)

      // Recurse to sub rows before aggregation
      groupedRows = Object.entries(groupedRows).map(
        ([groupByVal, subRows], index) => {
          let id = `${columnId}:${groupByVal}`
          id = parentId ? `${parentId}>${id}` : id

          subRows = groupRecursively(subRows, depth + 1, id)

          const values = aggregateRowsToValues(
            subRows,
            depth < existingGroupBy.length
          )

          const row = {
            id,
            isGrouped: true,
            groupByID: columnId,
            groupByVal,
            values,
            subRows,
            depth,
            index,
          }

          groupedFlatRows.push(row, ...subRows)

          return row
        }
      )

      return groupedRows
    }

    const groupedRows = groupRecursively(rows)

    // Assign the new data
    return [groupedRows, groupedFlatRows]
  }, [
    manualGroupBy,
    groupBy,
    rows,
    flatRows,
    flatColumns,
    userAggregations,
    groupByFn,
  ])

  const getAutoResetGroupBy = useGetLatest(autoResetGroupBy)

  useMountedLayoutEffect(() => {
    if (getAutoResetGroupBy()) {
      dispatch({ type: actions.resetGroupBy })
    }
  }, [dispatch, manaulGroupBy ? null : data])

  Object.assign(instance, {
    preGroupedRows: rows,
    preGroupedFlatRow: flatRows,
    groupedRows,
    groupedFlatRows,
    rows: groupedRows,
    flatRows: groupedFlatRows,
    toggleGroupBy,
  })
}
