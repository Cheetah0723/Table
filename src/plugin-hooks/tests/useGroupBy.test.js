import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { useTable } from '../../hooks/useTable'
import { useGroupBy } from '../useGroupBy'
import { useExpanded } from '../useExpanded'

const data = [
  {
    firstName: 'tanner',
    lastName: 'linsley',
    age: 29,
    visits: 100,
    status: 'In Relationship',
    progress: 50,
  },
  {
    firstName: 'derek',
    lastName: 'perkins',
    age: 40,
    visits: 40,
    status: 'Single',
    progress: 80,
  },
  {
    firstName: 'joe',
    lastName: 'bergevin',
    age: 45,
    visits: 20,
    status: 'Complicated',
    progress: 10,
  },
  {
    firstName: 'jaylen',
    lastName: 'linsley',
    age: 26,
    visits: 99,
    status: 'In Relationship',
    progress: 70,
  },
]

const defaultColumn = {
  Cell: ({ cell: { value }, column: { id } }) => `${id}: ${value}`,
  Filter: ({ filterValue, setFilter }) => (
    <input
      value={filterValue || ''}
      onChange={e => {
        setFilter(e.target.value || undefined) // Set undefined to remove the filter entirely
      }}
      placeholder="Search..."
    />
  ),
}

function Table({ columns, data }) {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
  } = useTable(
    {
      columns,
      data,
      defaultColumn,
    },
    useGroupBy,
    useExpanded
  )

  return (
    <table {...getTableProps()}>
      <thead>
        {headerGroups.map(headerGroup => (
          <tr {...headerGroup.getHeaderGroupProps()}>
            {headerGroup.headers.map(column => (
              <th {...column.getHeaderProps()}>
                {column.canGroupBy ? (
                  // If the column can be grouped, let's add a toggle
                  <span {...column.getGroupByToggleProps()}>
                    {column.isGrouped ? 'ðŸ›‘' : 'ðŸ‘Š'}
                  </span>
                ) : null}
                {column.render('Header')}
              </th>
            ))}
          </tr>
        ))}
      </thead>
      <tbody {...getTableBodyProps()}>
        {rows.map(
          (row, i) =>
            prepareRow(row) || (
              <tr {...row.getRowProps()}>
                {row.cells.map(cell => {
                  return (
                    <td {...cell.getCellProps()}>
                      {cell.isGrouped ? (
                        <>
                          <span
                            style={{
                              cursor: 'pointer',
                            }}
                            onClick={() => row.toggleExpanded()}
                          >
                            {row.isExpanded ? 'ðŸ‘‡' : 'ðŸ‘‰'}
                          </span>
                          {cell.render('Cell')} ({row.subRows.length})
                        </>
                      ) : cell.isAggregated ? (
                        cell.render('Aggregated')
                      ) : cell.isRepeatedValue ? null : (
                        cell.render('Cell')
                      )}
                    </td>
                  )
                })}
              </tr>
            )
        )}
      </tbody>
    </table>
  )
}

function roundedMedian(values) {
  let min = values[0] || ''
  let max = values[0] || ''

  values.forEach(value => {
    min = Math.min(min, value)
    max = Math.max(max, value)
  })

  return Math.round((min + max) / 2)
}

function App() {
  const columns = React.useMemo(
    () => [
      {
        Header: 'Name',
        columns: [
          {
            Header: 'First Name',
            accessor: 'firstName',
            aggregate: ['sum', 'count'],
            Aggregated: ({ cell: { value } }) => `${value} Names`,
          },
          {
            Header: 'Last Name',
            accessor: 'lastName',
            aggregate: ['sum', 'uniqueCount'],
            Aggregated: ({ cell: { value } }) => `${value} Unique Names`,
          },
        ],
      },
      {
        Header: 'Info',
        columns: [
          {
            Header: 'Age',
            accessor: 'age',
            aggregate: 'average',
            Aggregated: ({ cell: { value } }) => `${value} (avg)`,
          },
          {
            Header: 'Visits',
            accessor: 'visits',
            aggregate: 'sum',
            Aggregated: ({ cell: { value } }) => `${value} (total)`,
          },
          {
            Header: 'Status',
            accessor: 'status',
          },
          {
            Header: 'Profile Progress',
            accessor: 'progress',
            aggregate: roundedMedian,
            Aggregated: ({ cell: { value } }) => `${value} (med)`,
          },
        ],
      },
    ],
    []
  )

  return <Table columns={columns} data={data} />
}

test('renders a groupable table', () => {
  const { getAllByText, asFragment } = render(<App />)

  const groupByButtons = getAllByText('ðŸ‘Š')

  const beforeGrouping = asFragment()

  fireEvent.click(groupByButtons[1])

  const afterGrouping1 = asFragment()

  fireEvent.click(groupByButtons[3])

  const afterGrouping2 = asFragment()

  expect(beforeGrouping).toMatchDiffSnapshot(afterGrouping1)
  expect(afterGrouping1).toMatchDiffSnapshot(afterGrouping2)
})
