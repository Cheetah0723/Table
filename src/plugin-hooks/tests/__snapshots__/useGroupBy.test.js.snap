// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a groupable table 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -29,13 +29,13 @@
            <span
              class=""
              style="cursor: pointer;"
              title="Toggle GroupBy"
            >
-             ðŸ‘Š
+             ðŸ›‘
            </span>
-           First Name
+           Last Name
          </th>
          <th
            class=""
            colspan="1"
          >
@@ -44,11 +44,11 @@
              style="cursor: pointer;"
              title="Toggle GroupBy"
            >
              ðŸ‘Š
            </span>
-           Last Name
+           First Name
          </th>
          <th
            class=""
            colspan="1"
          >
@@ -109,138 +109,119 @@
          class=""
        >
          <td
            class=""
          >
-           firstName: tanner
-         </td>
-         <td
-           class=""
+           <span
+             style="cursor: pointer;"
            >
-           lastName: linsley
-         </td>
-         <td
-           class=""
-         >
-           age: 29
-         </td>
-         <td
-           class=""
-         >
-           visits: 100
-         </td>
-         <td
-           class=""
-         >
-           status: In Relationship
+             ðŸ‘‰
+           </span>
+           lastName: linsley (2)
          </td>
          <td
            class=""
          >
-           progress: 50
-         </td>
-       </tr>
-       <tr
-         class=""
-       >
-         <td
-           class=""
-         >
-           firstName: derek
-         </td>
-         <td
-           class=""
-         >
-           lastName: perkins
+           2 Names
          </td>
          <td
            class=""
          >
-           age: 40
+           27.5 (avg)
          </td>
          <td
            class=""
          >
-           visits: 40
+           199 (total)
          </td>
          <td
            class=""
          >
-           status: Single
+           status: null
          </td>
          <td
            class=""
          >
-           progress: 80
+           60 (med)
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class=""
          >
-           firstName: joe
+           <span
+             style="cursor: pointer;"
+           >
+             ðŸ‘‰
+           </span>
+           lastName: perkins (1)
          </td>
          <td
            class=""
          >
-           lastName: bergevin
+           1 Names
          </td>
          <td
            class=""
          >
-           age: 45
+           40 (avg)
          </td>
          <td
            class=""
          >
-           visits: 20
+           40 (total)
          </td>
          <td
            class=""
          >
-           status: Complicated
+           status: null
          </td>
          <td
            class=""
          >
-           progress: 10
+           80 (med)
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class=""
          >
-           firstName: jaylen
+           <span
+             style="cursor: pointer;"
+           >
+             ðŸ‘‰
+           </span>
+           lastName: bergevin (1)
          </td>
          <td
            class=""
          >
-           lastName: linsley
+           1 Names
          </td>
          <td
            class=""
          >
-           age: 26
+           45 (avg)
          </td>
          <td
            class=""
          >
-           visits: 99
+           20 (total)
          </td>
          <td
            class=""
          >
-           status: In Relationship
+           status: null
          </td>
          <td
            class=""
          >
-           progress: 70
+           10 (med)
          </td>
        </tr>
      </tbody>
    </table>
  </DocumentFragment>
`;

exports[`renders a groupable table 2`] = `
Snapshot Diff:
- First value
+ Second value

@@ -6,17 +6,29 @@
        <tr
          class=""
        >
          <th
            class=""
-           colspan="2"
+           colspan="1"
          >
            Name
          </th>
          <th
            class=""
-           colspan="4"
+           colspan="1"
+         >
+           Info
+         </th>
+         <th
+           class=""
+           colspan="1"
+         >
+           Name
+         </th>
+         <th
+           class=""
+           colspan="3"
          >
            Info
          </th>
        </tr>
        <tr
@@ -42,13 +54,13 @@
            <span
              class=""
              style="cursor: pointer;"
              title="Toggle GroupBy"
            >
-             ðŸ‘Š
+             ðŸ›‘
            </span>
-           First Name
+           Visits
          </th>
          <th
            class=""
            colspan="1"
          >
@@ -57,11 +69,11 @@
              style="cursor: pointer;"
              title="Toggle GroupBy"
            >
              ðŸ‘Š
            </span>
-           Age
+           First Name
          </th>
          <th
            class=""
            colspan="1"
          >
@@ -70,11 +82,11 @@
              style="cursor: pointer;"
              title="Toggle GroupBy"
            >
              ðŸ‘Š
            </span>
-           Visits
+           Age
          </th>
          <th
            class=""
            colspan="1"
          >
@@ -116,10 +128,13 @@
            >
              ðŸ‘‰
            </span>
            lastName: linsley (2)
          </td>
+         <td
+           class=""
+         />
          <td
            class=""
          >
            2 Names
          </td>
@@ -129,15 +144,10 @@
            27.5 (avg)
          </td>
          <td
            class=""
          >
-           199 (total)
-         </td>
-         <td
-           class=""
-         >
            status: null
          </td>
          <td
            class=""
          >
@@ -157,22 +167,20 @@
            </span>
            lastName: perkins (1)
          </td>
          <td
            class=""
-         >
-           1 Names
-         </td>
+         />
          <td
            class=""
          >
-           40 (avg)
+           1 Names
          </td>
          <td
            class=""
          >
-           40 (total)
+           40 (avg)
          </td>
          <td
            class=""
          >
            status: null
@@ -196,22 +204,20 @@
            </span>
            lastName: bergevin (1)
          </td>
          <td
            class=""
-         >
-           1 Names
-         </td>
+         />
          <td
            class=""
          >
-           45 (avg)
+           1 Names
          </td>
          <td
            class=""
          >
-           20 (total)
+           45 (avg)
          </td>
          <td
            class=""
          >
            status: null
`;
